daemon/containerd/image_delete.go

```diff
@@ -59,6 +75,52 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
 		return nil, err
 	}
 
+	imgID := image.ID(img.Target.Digest)
+
+	if isImageIDPrefix(imgID.String(), imageRef) {
+		return i.deleteAll(ctx, img, force, prune)
+	}
+
+	if !i.isSingleReference(ctx, imgID) {
```

Comment:
I'm a little worried that "there is only one reference to `imgID` implies that the reference is `img`" is a leap of logic which is not necessarily true, especially if there are multiple concurrent image-store operations going on. It seems like the question is "are there any images other than `img` which reference the same content as `img`," which could be answered more precisely by checking whether the set of images which target `img.Target.Digest` and are neither named `img.Name` nor are dangling is the empty set.

Reply:
I'm not sure I quite understand your point, but maybe you can clarify – to get things in clear terms, lets reformulate: instead of "there is only one reference to `imgID` implies that the reference is `img`" we can use "I have a reference `img` and the set of images which target `img.Target.Digest` that are not dangling has cardinality 1`", which is more concretely the test the code is doing.

It seems to me that
 ```
I have a reference `img` and the set of images which target `img.Target.Digest` that are not dangling has cardinality 1`
```
and 
```
The set of images which target `img.Target.Digest` and are neither named `img.Name` nor are dangling is the empty set
```
are equivalent statements (taking into account that `img.Name` has to be unique in the containerd store).

Reply:
Writing this I think I've realized the benefit of the 2nd approach is that we can do this check in a more "atomic" way 😅 Thanks, I'll make those changes :)